File: moling-server/app/dao/phase4_dao.py
"""墨灵 (Moling) — Phase 4 Task DAO."""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.dao.base_dao import BaseDAO
from app.models.phase4_task import Phase4Task


class Phase4DAO(BaseDAO[Phase4Task]):
    """Data access object for Phase4Task model."""

    def __init__(self):
        super().__init__(Phase4Task)

    async def get_by_nonce(
        self,
        db: AsyncSession,
        nonce: str,
    ) -> Phase4Task | None:
        """Get task by nonce (for idempotency check)."""
        stmt = select(Phase4Task).where(Phase4Task.nonce == nonce)
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def get_by_chapter(
        self,
        db: AsyncSession,
        chapter_id: str,
        *,
        status: str | None = None,
    ) -> list[Phase4Task]:
        """Get tasks by chapter ID."""
        stmt = select(Phase4Task).where(Phase4Task.chapter_id == chapter_id)
        
        if status:
            stmt = stmt.where(Phase4Task.status == status)
        
        stmt = stmt.order_by(Phase4Task.created_at.desc())
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def get_by_project(
        self,
        db: AsyncSession,
        project_id: str,
        *,
        status: str | None = None,
    ) -> list[Phase4Task]:
        """Get tasks by project ID."""
        stmt = select(Phase4Task).where(Phase4Task.project_id == project_id)
        
        if status:
            stmt = stmt.where(Phase4Task.status == status)
        
        stmt = stmt.order_by(Phase4Task.created_at.desc())
        result = await db.execute(stmt)
        return list(result.scalars().all())


# Singleton instance
phase4_dao = Phase4DAO()

File: moling-server-rs/crates/moling-db/src/dao/phase4_dao.rs
//! Phase4 DAO — Phase 4 task CRUD with idempotency support.

use moling_core::error::{AppError, AppResult};
use sea_orm::{ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, PaginatorTrait, QueryFilter, QueryOrder, QuerySelect};

use crate::entities::phase4_task::{self, ActiveModel, Entity as Phase4Task, Model as Phase4TaskModel};

/// Phase 4 task data access object.
#[derive(Clone, Default)]
pub struct Phase4Dao;

impl Phase4Dao {
    /// Find a task by ID.
    pub async fn find_by_id(
        &self,
        db: &DatabaseConnection,
        id: i32,
    ) -> AppResult<Option<Phase4TaskModel>> {
        Phase4Task::find_by_id(id)
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%id, "Phase4: database error finding by id: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Find a task by nonce (for idempotency check).
    pub async fn find_by_nonce(
        &self,
        db: &DatabaseConnection,
        nonce: &str,
    ) -> AppResult<Option<Phase4TaskModel>> {
        Phase4Task::find()
            .filter(phase4_task::Column::Nonce.eq(nonce))
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%nonce, "Phase4: database error finding by nonce: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List tasks by chapter ID, optionally filtered by state.
    pub async fn list_by_chapter(
        &self,
        db: &DatabaseConnection,
        chapter_id: &str,
        state: Option<&str>,
    ) -> AppResult<Vec<Phase4TaskModel>> {
        let mut query = Phase4Task::find()
            .filter(phase4_task::Column::ChapterId.eq(chapter_id));

        if let Some(s) = state {
            query = query.filter(phase4_task::Column::State.eq(s));
        }

        query
            .order_by_desc(phase4_task::Column::CreatedAt)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(%chapter_id, "Phase4: database error listing by chapter: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List tasks by project ID, optionally filtered by state.
    pub async fn list_by_project(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        state: Option<&str>,
    ) -> AppResult<Vec<Phase4TaskModel>> {
        let mut query = Phase4Task::find()
            .filter(phase4_task::Column::ProjectId.eq(project_id));

        if let Some(s) = state {
            query = query.filter(phase4_task::Column::State.eq(s));
        }

        query
            .order_by_desc(phase4_task::Column::CreatedAt)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(%project_id, "Phase4: database error listing by project: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List tasks by status with pagination.
    pub async fn list_by_status(
        &self,
        db: &DatabaseConnection,
        status: &str,
        skip: u64,
        limit: u64,
    ) -> AppResult<Vec<Phase4TaskModel>> {
        Phase4Task::find()
            .filter(phase4_task::Column::State.eq(status))
            .order_by_desc(phase4_task::Column::CreatedAt)
            .offset(skip)
            .limit(limit)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(%status, "Phase4: database error listing by status: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Count tasks matching a given state.
    pub async fn count_by_status(
        &self,
        db: &DatabaseConnection,
        status: &str,
    ) -> AppResult<u64> {
        Phase4Task::find()
            .filter(phase4_task::Column::State.eq(status))
            .count(db)
            .await
            .map_err(|e| {
                tracing::error!(%status, "Phase4: database error counting by status: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Create a new phase 4 task.
    pub async fn create(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<Phase4TaskModel> {
        model.insert(db).await.map_err(|e| {
            tracing::error!("Phase4: database error creating: {e}");
            AppError::internal("Database insert failed")
        })
    }

    /// Update an existing phase 4 task.
    pub async fn update(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<Phase4TaskModel> {
        model.update(db).await.map_err(|e| {
            tracing::error!("Phase4: database error updating: {e}");
            AppError::internal("Database update failed")
        })
    }
}
